addons/binding/org.openhab.binding.mihome/src/main/java/org/openhab/binding/mihome/handler/XiaomiDeviceBaseHandler.java

```diff
@@ -0,0 +1,237 @@
+ * @author Kuba Wolanin - Added voltage and low battery report
+ * @author Dimalo - Added cube rotation, heartbeat and voltage handling, configurable window and motion delay, Aqara
+ *         switches
+ */
+public abstract class XiaomiDeviceBaseHandler extends BaseThingHandler implements XiaomiItemUpdateListener {
+
+    public static final Set<ThingTypeUID> SUPPORTED_THING_TYPES = new HashSet<>(
+            Arrays.asList(THING_TYPE_GATEWAY, THING_TYPE_SENSOR_HT, THING_TYPE_SENSOR_MOTION, THING_TYPE_SENSOR_SWITCH,
+                    THING_TYPE_SENSOR_MAGNET, THING_TYPE_SENSOR_CUBE, THING_TYPE_SENSOR_AQARA1,
+                    THING_TYPE_SENSOR_AQARA2, THING_TYPE_ACTOR_AQARA1, THING_TYPE_ACTOR_AQARA2, THING_TYPE_ACTOR_PLUG));
+
+    private static final long ONLINE_TIMEOUT = 2 * 60 * 60 * 1000; // 2 hours
+
+    private JsonParser parser = new JsonParser();
+
+    private XiaomiBridgeHandler bridgeHandler;
+
+    String itemId;
+
+    Logger logger = LoggerFactory.getLogger(this.getClass().getName());
+
+    public XiaomiDeviceBaseHandler(Thing thing) {
+        super(thing);
+    }
+
+    @Override
+    public void initialize() {
+        final String configItemId = (String) getConfig().get(ITEM_ID);
+        if (configItemId != null) {
+            itemId = configItemId;
+        }
+        // schedule init with random delay between 200ms and 1sec
+        scheduler.schedule(new Runnable() {
+            @Override
+            public void run() {
+                updateThingStatus();
+            }
+        }, (int) Math.max((Math.random() * 1000), 200), TimeUnit.MILLISECONDS);
+    }
+
+    @Override
+    public void dispose() {
+        logger.debug("Handler disposes. Unregistering listener.");
+        if (itemId != null) {
+            XiaomiBridgeHandler bridgeHandler = getXiaomiBridgeHandler();
+            if (bridgeHandler != null) {
+                bridgeHandler.unregisterItemListener(this);
+                this.bridgeHandler = null;
+            }
+            itemId = null;
+        }
+    }
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        logger.debug("Device {} on channel {} received command {}", itemId, channelUID, command);
+        if (command.toString().toLowerCase().equals("refresh")) {
+            return;
+        }
+        execute(channelUID, command);
+    }
+
+    @Override
+    public void onItemUpdate(String sid, String command, JsonObject message) {
+        if (itemId != null && itemId.equals(sid)) {
+            logger.debug("Item got update: {}", message.toString());
+            try {
+                JsonObject data = parser.parse(message.get("data").getAsString()).getAsJsonObject();
+                parseCommand(command, data);
+            } catch (Exception e) {
+                logger.error("Unable to parse message {}", message);
+                logger.error("Caught Exeption {}", e);
+            }
+            updateThingStatus();
```

Comment:
this will update the Online status accordingly - so make it OFFLINE if it needs to. Receiving garbage does not mean everything is wrong. we are relying on multicast datagrams, hence UDP, which does not always deliver right?